From dc6ae822fa2c091833f8ffde7bfa4e0503543d2f Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Fri, 22 May 2020 14:40:39 +0200
Subject: [PATCH] nano: Integrate with Balena and load kernel from rootfs

This is a port to L4T 32.4.2 of the existent patches,
as jetson nano u-boot machine name was changed.

It integrates BalenaOS with Jetson Nano's u-boot.

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 configs/p3450-0000_defconfig    |  7 ++++++-
 include/config_distro_bootcmd.h | 14 ++++++++++----
 include/configs/p3450-0000.h    |  2 ++
 3 files changed, 18 insertions(+), 5 deletions(-)

diff --git a/configs/p3450-0000_defconfig b/configs/p3450-0000_defconfig
index f78e1d3867..179de7890d 100644
--- a/configs/p3450-0000_defconfig
+++ b/configs/p3450-0000_defconfig
@@ -20,7 +20,7 @@ CONFIG_CMD_SF=y
 CONFIG_CMD_SPI=y
 CONFIG_CMD_USB=y
 CONFIG_CMD_USB_MASS_STORAGE=y
-# CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_SETEXPR=y
 # CONFIG_CMD_NFS is not set
 CONFIG_CMD_EXT4_WRITE=y
 CONFIG_OF_LIVE=y
@@ -62,3 +62,8 @@ CONFIG_ENV_OFFSET=0xFFFFE000
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_POSITION_INDEPENDENT=y
 CONFIG_DISABLE_SDMMC1_EARLY=y
+CONFIG_CMD_PART=y
+CONFIG_PARTITION_UUIDS=y
+CONFIG_CMD_IMPORTENV=y
+CONFIG_BOOTDELAY=0
+CONFIG_CMD_EXPORTENV=y
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index fc0935fa21..98728e3537 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -25,6 +25,8 @@
  * message that includes some other pre-processor symbols in the text.
  */
 
+#define BOOT_TARGET_DEFAULTDEVPLIST    "C"
+
 #define BOOTENV_SHARED_BLKDEV_BODY(devtypel) \
 		"if " #devtypel " dev ${devnum}; then " \
 			"devtype=" #devtypel "; " \
@@ -437,7 +439,7 @@
 		"fi\0"                                                    \
 	\
 	"boot_a_script="                                                  \
-		"load ${devtype} ${devnum}:${distro_bootpart} "           \
+               "load ${resin_dev_type} ${resin_dev_index}:${resin_boot_part} "\
 			"${scriptaddr} ${prefix}${script}; "              \
 		"source ${scriptaddr}\0"                                  \
 	\
@@ -464,7 +466,8 @@
 		"\0"                                                      \
 	\
 	"scan_dev_for_boot_part="                                         \
-		"part list ${devtype} ${devnum} -bootable devplist; "     \
+               "setenv devplist $resin_root_part; "                      \
+               "env exists devplist || part list ${resin_dev_type} ${resin_dev_index} -bootable devplist; "     \
 		"env exists devplist || setenv devplist 1; "              \
 		"for distro_bootpart in ${devplist}; do "                 \
 			"if fstype ${devtype} "                           \
@@ -481,10 +484,13 @@
 		BOOTENV_SET_NVME_NEED_INIT                                \
 		BOOTENV_SET_IDE_NEED_INIT                                 \
 		BOOTENV_SET_VIRTIO_NEED_INIT                              \
+		"setenv resin_kernel_load_addr ${loadaddr}; "     \
 		"for target in ${boot_targets}; do "                      \
-			"run bootcmd_${target}; "                         \
+			"if env exist resin_set_kernel_root; then "       \
+				"run resin_set_kernel_root; run set_os_cmdline; "  \
+				"run bootcmd_${target}; "                 \
+			"fi; "                                            \
 		"done\0"
-
 #ifndef CONFIG_BOOTCOMMAND
 #define CONFIG_BOOTCOMMAND "run distro_bootcmd"
 #endif
diff --git a/include/configs/p3450-0000.h b/include/configs/p3450-0000.h
index eff463eecb..d490e76a92 100644
--- a/include/configs/p3450-0000.h
+++ b/include/configs/p3450-0000.h
@@ -17,6 +17,7 @@
 #define CONFIG_TEGRA_BOARD_STRING	"NVIDIA P3450-0002"
 #define BOOT_TARGET_DEVICES(func) \
 	func(MMC, mmc, 0) \
+	func(MMC, mmc, 1) \
 	func(PXE, pxe, na) \
 	func(DHCP, dhcp, na)
 #define CONFIG_SYS_MMC_ENV_DEV		0
@@ -29,6 +30,7 @@
 
 /* Only MMC/PXE/DHCP for now, add USB back in later when supported */
 #define BOOT_TARGET_DEVICES(func) \
+	func(MMC, mmc, 0) \
 	func(MMC, mmc, 1) \
 	func(PXE, pxe, na) \
 	func(DHCP, dhcp, na)
-- 
2.17.1

